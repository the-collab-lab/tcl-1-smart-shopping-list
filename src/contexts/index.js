// NOTE: This file (like the index.js files throughout the app)
// is a way to "bundle" our imports into a more consistent format
// and to access them from a more consistent location. For instance,
// the export "bundle" statement below allows us to write this single-
// line import statement format:
//   import { TokenProvider, TokenContext } from './path/to/this/directory'
// instead of this multi-line format:
//   import { TokenProvider } from './path/to/this/directory/token.context;
//   import { ListProvider } from './path/to/this/directory/list.context;
//
// In smaller apps like this it's likely an overengineering of imports,
// but it helps keep things tidy, which helps anyone who's brain is easily
// confused by clutter.
//
// Additionally, if you move these imported files around (ie. add nested
// directories, add additional contexts that you need access to), updating
// this file to point to where the file "lives" after the update is the only
// place in the codebase where you should have to update the file path in
// order to find the required imports.

import { TokenContext, TokenProvider } from './token.context';
import { ListContext, ListProvider } from './list.context';

export { TokenContext, ListContext, TokenProvider, ListProvider };
