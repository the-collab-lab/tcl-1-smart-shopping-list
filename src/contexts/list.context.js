import React, { createContext, useState } from 'react';
import PropTypes from 'prop-types';

const ListContext = createContext({
  value: [],
  updateListValue: () => {},
});

const ListProvider = ({ children }) => {
  // NOTE: setListValue is just a wrapper around our setList hook action with some helper
  // logic included so that it formats or otherwise handles the data before the list
  // state's value is updated.
  //
  // In this particular helper we are
  // 1) taking the argument to the setList, which is a record of the previous value for
  //    `list` that will be overwritten as soon as our setList hook action is called, we
  //    are going to merge the two object with the javascript spread syntax. In short,
  //    we can merge these two objects:
  //      const objectA = { greeting: 'hello', action: 'wave' };
  //      const objectB = { greeting: 'goodbye' };
  //    into one object like this:
  //      const objectC = { ...objectA, ...objectB };
  //    which results in the following object:
  //      console.log(objectC) ==> { greeting: 'goodbye', action: 'wave' }
  //
  // 2) once we have that newly-merged object, we return the new object, which
  //    in this case passes it to setList, which updates our fancy ContextProvider
  //    with a fresh value without overwriting all the other functions it contains.
  //
  // NOTE: in case the return statement looks weird, these two variable declarations
  // result in the same object, just different (ES6 / ECMAScript 2015?) syntax:
  //   const objectD = { value }
  //   const objectE = { value: value }
  const setListValue = newList => {
    setList(prev => {
      const prevList = { ...prev };
      return { ...prevList, ...{ list: newList } };
    });
  };

  // NOTE: this is where we're passing the init/reset values to our fancy
  // new ListProvider "wrapper" component
  const listValueReset = { list: [], setListValue };

  // NOTE: under the hood it's really just a useState hook  with some
  // extra features managing the state of our application.
  const [list, setList] = useState(listValueReset);

  // NOTE: Here you can see that behind our concisely named provider component
  // (<ListContext /> in this case) is the stock ListContext Provider syntax
  // the usefulness here is that we're able to pull all the context logic out
  // into it's own component(s) in order to keep app.js more tidy and concise.
  return <ListContext.Provider value={list}>{children}</ListContext.Provider>;
};

// NOTE: whichever (single) portion of this file you "export default" will be
// accessible from other files with the following import statement format:
//   import MagicComponent from './path/to/this/directory/this.file.js'
// whereas whatever (allows for more than one!) portion of this file you "export"
// is accessible from other files with the following import statement format:
//   import { MagicComponent, SpecialComponent } from './path/to/this/directory/this.file.js
//
// Additionally, do note that not every part of your file needs to or should be
// exported. There will likely be many instances where you keep single-use helpers
// in the file with where they'll be used, but since they aren't applicable to
// other areas of our application, it's fine to leave them tucked away without an export.
export { ListContext, ListProvider };

ListProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
ListProvider.defaultProps = {};
