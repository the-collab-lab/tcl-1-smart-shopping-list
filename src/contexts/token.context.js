import React, { createContext, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import getToken from '../lib/token';

const TokenContext = createContext({
  token: '',
  setTokenValue: () => {},
});

const TokenProvider = ({ children }) => {
  const setTokenValue = newToken => {
    setToken(prev => {
      const prevToken = { ...prev };
      const merged = { ...prevToken, ...{ token: newToken } };

      localStorage.setItem('token', newToken);
      return merged;
    });
  };

  const confirmToken = () => {
    const storedToken = localStorage.getItem('token');

    if (!!storedToken) return storedToken;
    else return getToken();
  };

  // NOTE: this is where we're passing the init/reset values to our fancy
  // new TokenProvider "wrapper" component
  const tokenValueReset = {
    // NOTE ON TOKEN VALUE: since we store tokens in localStorage so that
    // they persist between reloads and different sessions (on the same
    // device/browser), we'd like to reuse a stored token if it exists,
    // otherwise set an init value
    token: localStorage.getItem('token') || '',
    setTokenValue,
    confirmToken,
  };

  // NOTE: under the hood it's really just a useState hook  with some
  // extra features managing the state of our application.
  const [token, setToken] = useState(tokenValueReset);

  // NOTE: Here you can see that behind our concisely named provider component
  // (<TokenContext /> in this case) is the stock TokenContext Provider syntax
  // the usefulness here is that we're able to pull all the context logic out
  // into it's own component(s) in order to keep app.js more tidy and concise.
  return (
    <TokenContext.Provider value={token}>{children}</TokenContext.Provider>
  );
};

// NOTE: whichever (single) portion of this file you "export default" will be
// accessible from other files with the following import statement format:
//   import MagicComponent from './path/to/this/directory/this.file.js'
// whereas whatever (allows for more than one!) portion of this file you "export"
// is accessible from other files with the following import statement format:
//   import { MagicComponent, SpecialComponent } from './path/to/this/directory/this.file.js
//
// Additionally, do note that not every part of your file needs to or should be
// exported. There will likely be many instances where you keep single-use helpers
// in the file with where they'll be used, but since they aren't applicable to
// other areas of our application, it's fine to leave them tucked away without an export.
export { TokenContext, TokenProvider };

TokenProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
TokenProvider.defaultProps = {};
