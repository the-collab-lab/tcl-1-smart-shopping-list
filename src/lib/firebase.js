// NOTE: import only the Firebase modules that you need in your app... except
// for the second line, which makes both the linter and react-firebase happy
import firebase from 'firebase/compat/app';
import 'firebase/compat/firestore';

// Initalize Firebase.
const config = {
    apiKey: "AIzaSyDvVB6SIAhXqa3chcJgaJi_XGkd1W3xIhI",
    authDomain: "tcl-1-smart-shopping-list.firebaseapp.com",
    projectId: "tcl-1-smart-shopping-list",
    storageBucket: "tcl-1-smart-shopping-list.appspot.com",
    messagingSenderId: "490626563800",
    appId: "1:490626563800:web:1c6c138d68d09a6f8de1ad"
};

let fb = firebase.initializeApp(config);

export { fb };
