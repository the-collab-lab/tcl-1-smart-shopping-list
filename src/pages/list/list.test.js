import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter } from 'react-router-dom';
import TestRenderer from 'react-test-renderer';
import { fb as firebase } from '../../lib/firebase.js';
import { FirestoreProvider } from 'react-firestore';
import List from './index';
import { TokenProvider, ListProvider } from './../../contexts';

// NOTE: programatically "pushing" to our app's routing history
// means have to mock the history object and it's push function

// TODO: mock additional context values and expect the same tests to
// pass if it CAN see some sort of stored token -- see ListTest for an example

describe('list is only provided list context values, redirects', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <ListProvider value={{ list: [] }}>
        <FirestoreProvider firebase={firebase}>
          <BrowserRouter>
            <List history={{ push: jest.fn() }} />
          </BrowserRouter>
        </FirestoreProvider>
      </ListProvider>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it('snapshot on load', () => {
    const tree = TestRenderer.create(
      <ListProvider value={{ list: [] }}>
        <FirestoreProvider firebase={firebase}>
          <BrowserRouter>
            <List history={{ push: jest.fn() }} />
          </BrowserRouter>
        </FirestoreProvider>
      </ListProvider>
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});

describe('list DOES NOT contain items', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <TokenProvider value={{ token: 'abc123' }}>
        <ListProvider value={{ list: [] }}>
          <FirestoreProvider firebase={firebase}>
            <BrowserRouter>
              <List history={{ push: jest.fn() }} />
            </BrowserRouter>
          </FirestoreProvider>
        </ListProvider>
      </TokenProvider>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it('snapshot on load', () => {
    const tree = TestRenderer.create(
      <TokenProvider value={{ token: 'abc123' }}>
        <ListProvider value={{ list: [] }}>
          <FirestoreProvider firebase={firebase}>
            <BrowserRouter>
              <List history={{ push: jest.fn() }} />
            </BrowserRouter>
          </FirestoreProvider>
        </ListProvider>
      </TokenProvider>
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});

describe('list DOES contain items', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      // <TokenContext.Provider value={{ token: 'abc123' }}>
      <ListProvider value={{ list: [] }}>
        <FirestoreProvider firebase={firebase}>
          <BrowserRouter>
            <List history={{ push: jest.fn() }} />
          </BrowserRouter>
        </FirestoreProvider>
      </ListProvider>,
      // </TokenContext.Provider>
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it('snapshot on load', () => {
    const list = [
      { name: 'one', duration: 'soon', listToken: 'abc123' },
      { name: 'two', duration: 'soon', listToken: 'abc123' },
      { name: 'three', duration: 'soon', listToken: 'abc123' },
    ];

    const tree = TestRenderer.create(
      // <TokenContext.Provider value={{ token: 'abc123' }}>
      <ListProvider value={{ list }}>
        <FirestoreProvider firebase={firebase}>
          <BrowserRouter>
            <List history={{ push: jest.fn() }} />
          </BrowserRouter>
        </FirestoreProvider>
      </ListProvider>
      // </TokenContext.Provider>
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
