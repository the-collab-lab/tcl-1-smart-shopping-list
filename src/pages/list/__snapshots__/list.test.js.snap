// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list DOES NOT contain items snapshot on load 1`] = `
<main
  className="page-wrapper css-vdsylo"
>
  <header
    className="header css-527nkd"
  >
    <a
      className="back-link css-nuywz6"
      href="/"
      onClick={[Function]}
      visualstate="hidden"
    >
      Go Back
    </a>
    <a
      className="title-link css-gq9b4g"
      href="/"
      onClick={[Function]}
      visualstate="default"
    >
      <h1
        className="css-r73csp"
      >
        iNeedToBuy
      </h1>
    </a>
    <a
      className="example-link css-gq9b4g"
      href="http://app.ineedtobuy.xyz/"
      rel="no-referrer"
      visualstate="default"
    >
      See Example
    </a>
  </header>
  <section
    className="content-wrapper css-1xpf03y"
  >
    <p
      className="loading"
    >
      Loading...
    </p>
    <footer
      className="footer css-rxnh45"
    >
      <section>
        <p>
          No purchases just yet.
        </p>
        <a
          href="/add-item"
          onClick={[Function]}
        >
          Add Item
        </a>
      </section>
    </footer>
  </section>
  <footer
    className="footer css-1i3mvty"
  >
    <a
      className="list-link css-gq9b4g"
      href="/"
      onClick={[Function]}
      visualstate="default"
    >
      My List
    </a>
    <a
      className="add-item-link css-gq9b4g"
      href="/add-item"
      onClick={[Function]}
      visualstate="default"
    >
      Add Item
    </a>
  </footer>
</main>
`;

exports[`list DOES contain items snapshot on load 1`] = `
<main
  className="page-wrapper css-vdsylo"
>
  <header
    className="header css-527nkd"
  >
    <a
      className="back-link css-nuywz6"
      href="/"
      onClick={[Function]}
      visualstate="hidden"
    >
      Go Back
    </a>
    <a
      className="title-link css-gq9b4g"
      href="/"
      onClick={[Function]}
      visualstate="default"
    >
      <h1
        className="css-r73csp"
      >
        iNeedToBuy
      </h1>
    </a>
    <a
      className="example-link css-gq9b4g"
      href="http://app.ineedtobuy.xyz/"
      rel="no-referrer"
      visualstate="default"
    >
      See Example
    </a>
  </header>
  <section
    className="content-wrapper css-1xpf03y"
  >
    <p
      className="loading"
    >
      Loading...
    </p>
    <footer
      className="footer css-rxnh45"
    >
      <section>
        <p>
          No purchases just yet.
        </p>
        <a
          href="/add-item"
          onClick={[Function]}
        >
          Add Item
        </a>
      </section>
    </footer>
  </section>
  <footer
    className="footer css-1i3mvty"
  >
    <a
      className="list-link css-gq9b4g"
      href="/"
      onClick={[Function]}
      visualstate="default"
    >
      My List
    </a>
    <a
      className="add-item-link css-gq9b4g"
      href="/add-item"
      onClick={[Function]}
      visualstate="default"
    >
      Add Item
    </a>
  </footer>
</main>
`;

exports[`list is only provided list context values, redirects snapshot on load 1`] = `
<main
  className="page-wrapper css-vdsylo"
>
  <header
    className="header css-527nkd"
  >
    <a
      className="back-link css-nuywz6"
      href="/"
      onClick={[Function]}
      visualstate="hidden"
    >
      Go Back
    </a>
    <a
      className="title-link css-gq9b4g"
      href="/"
      onClick={[Function]}
      visualstate="default"
    >
      <h1
        className="css-r73csp"
      >
        iNeedToBuy
      </h1>
    </a>
    <a
      className="example-link css-gq9b4g"
      href="http://app.ineedtobuy.xyz/"
      rel="no-referrer"
      visualstate="default"
    >
      See Example
    </a>
  </header>
  <section
    className="content-wrapper css-1xpf03y"
  >
    <p
      className="loading"
    >
      Loading...
    </p>
    <footer
      className="footer css-rxnh45"
    >
      <section>
        <p>
          No purchases just yet.
        </p>
        <a
          href="/add-item"
          onClick={[Function]}
        >
          Add Item
        </a>
      </section>
    </footer>
  </section>
  <footer
    className="footer css-1i3mvty"
  >
    <a
      className="list-link css-gq9b4g"
      href="/"
      onClick={[Function]}
      visualstate="default"
    >
      My List
    </a>
    <a
      className="add-item-link css-gq9b4g"
      href="/add-item"
      onClick={[Function]}
      visualstate="default"
    >
      Add Item
    </a>
  </footer>
</main>
`;
